src/include/sof/lib/ams.h

```diff
@@ -0,0 +1,295 @@
+#define AMS_INVALID_SLOT 0xFF
+/* Wildcard for module_id and instance_id values */
+#define AMS_ANY_ID 0xFFFF
+
+/* Size of slots message, module id and instance id */
+#define AMS_SLOT_SIZE(msg) (AMS_MESSAGE_SIZE(msg) + sizeof(uint16_t) * 2)
+#define AMS_MESSAGE_SIZE(msg) (sizeof(*msg) - sizeof(char) + (sizeof(char) * (msg->message_length)))
+
+/**
+ * \brief IXC message payload
+ *
+ * ams_message_payload - contains the actual Async Msg payload
+ */
+struct ams_message_payload {
+	/* Message IDs are assigned dynamically on new message entry creation
+	   For a new payload should be acquired by ams_get_message_type_id */
+	uint32_t message_type_id;
+	/* Producers module ID */
+	uint16_t producer_module_id;
+	/* Producers instance ID */
+	uint16_t producer_instance_id;
+	/* Message length */
+	uint32_t message_length;
+	/* Message payload */
+	uint8_t *message;
+};
+
+struct ams_slot {
+	uint16_t module_id;
+	uint16_t instance_id;
+	union {
+		struct ams_message_payload msg;
+		uint8_t msg_raw[AMS_MAX_MSG_SIZE];
+	} u;
+	uint32_t __aligned(PLATFORM_DCACHE_ALIGN) pad[0];
+};
+
+/**
+ * \brief ams_msg_callback_fn
+ *
+ * Each subscriber provides this handler function for each message ID
+ */
+typedef void (*ams_msg_callback_fn)(const struct ams_message_payload *const ams_message_payload,
+				    void *ctx);
+
+
+/**
+ * \brief Internal struct ams_consumer_entry
+ *
+ * Describes a single consumer's subscription to a single message.
+ * Array of 'ams_consumer_entry' structs forms AsyncMessageService's routing
+ * table which allows for message dispatch.
+ */
+struct ams_consumer_entry {
+	/* Message ID that will be routed via this entry */
+	uint32_t message_type_id;
+	/* Callback provided by the subscribed consumer */
+	ams_msg_callback_fn consumer_callback;
+	/* Additional context for consumer_callback (optional) */
+	void *ctx;
+	/* Subscribed consumer's Module ID */
+	uint16_t consumer_module_id;
+	/* Subscribed consumer's Module Instance ID */
+	uint8_t consumer_instance_id;
+	/* Subscribed consumer's Module core id. Saved to speed up routing */
+	uint8_t consumer_core_id;
+};
+
+struct ams_producer {
+	/* Message ID that will be routed via this entry */
+	uint32_t message_type_id;
+	/* Subscribed producer's Module ID */
+	uint16_t producer_module_id;
+	/* Subscribed producer's Module Instance ID */
+	uint8_t producer_instance_id;
+};
+
+struct uuid_idx {
+	uint32_t message_type_id;
+	uint8_t message_uuid[UUID_SIZE];
+};
+
+struct ams_shared_context {
+	/* should be only used with ams_acquire/release function, not generic ones */
+	struct coherent c;
+
+	uint32_t last_used_msg_id;
+	struct ams_consumer_entry rt_table[AMS_ROUTING_TABLE_SIZE];
+	struct ams_producer producer_table[AMS_ROUTING_TABLE_SIZE];
+	struct uuid_idx uuid_table[AMS_SERVICE_UUID_TABLE_SIZE];
+
+	uint32_t slot_uses[CONFIG_CORE_COUNT];
+	/* marks which core already processed slot */
+	uint32_t slot_done[CONFIG_CORE_COUNT];
+
+	struct ams_slot slots[CONFIG_CORE_COUNT];
```

Comment:
So, slots aren't per-core, they are per message, right? Every new message takes a new slot with various descriptors, e.g. masks, describing to which cores those messages have to be delivered. Then why is the slot number equal to the core number?

Reply:
> There are expected flows described in the internal intel docs.

`src/lib/ams.c` is not located in an Intel-specific directory. Should it?

Reply:
This is a purely software feature, not hardware dependent. Should it?
Every platform can use it, if it were platform specific, than the code would be duplicated per platform. AFAIK it is agreed that sof is moving towards async messages instead of notifications.

Reply:
> Every platform can use it

If it's meant as not specific to Intel then comments, discussions and explanations cannot rely or depend on internal documents.

Reply:
@lyakh 
1. Yes. At least it's not explicitly. I could think of a scenario, when 2 messages are sent to 1 core only and the slots get queued. Since slots are per message, you could have 2 separate messages in separate slots targeting the same core if there is a slot left and the first message is not yet processed. This should be handled.
2. yes
3. Access conflicts are handled through coherent api and slots are in shared context. If no slots are free, than it's producers duty to handle blocked scenarios (ie. retry on failed send). Messages are one-way only, so producer is only concerned about sending, not handling.
4.  There is a bitmask of pending slots in ams_task. So on receiving idc message, the core sets respective tasks pending_slots bits accordingly to the slots it should handle. There's still need to have a shared memory copy of the message payload, as the messages are sent asychronously. There is no guarantee, that the payload on producers side still exists when the receiver handles the msg.

@marc-hb sure, I was not trying to use it as an argument, but as an example for ease of explanation, since we both have access to this resource.

Reply:
> This of course assumes, that all callbacks are handled before next message is sent. In an unlikely event of multiple multicore messages being sent at the same time (meaning slots are not freed), an error is returned and is expected to be handled by the producer.

Can we please clarify and document this: either only one message can be in processing at any time, then we can simplify the implementation, or multiple messages can be queued, then we need to support that properly. If only one message can be handled, then an attempt to send another one while the previous one is still being processed should return an error immediately - that should be a part of the `message_send()` implementation.

Reply:
Why do you want to make those assumptions? Async messages are asynchronous and don't care about the received processing part. We have to satisfy the transport layer conditions: receiver exists, there is a free slot to pass the message. There's no need to consider what type of message it is or whether the message being sent is the same, but to a different core. Or if the previous one finished processing. Or if the processing is being handled properly on the receiving end. There's only the need to satisfy transport conditions. Do we have the receiver? Where is the receiver? Do we have a free slot to send this message to this core? If yes, no problem. If no, than return error to the producer.
Having multiple messages in slots at the same time may happen and is not necessarily a bad thing, as again - the sending transport part does not care about the message type, only routing.
Let's look at an example.
Two producers send given messages: A to cores 1 and 2 from core 0 and B to core 1 from core 0

1. Message A is stored in slot 0, routed to core 1
2. Message A is stored in slot 1, routed to core 2
Something takes too long (assumption is no heavy processing is done in context of Asynchronous Message), Message A is not processed on any of the cores yet.
4. Message B is stored in slot 2, routed to core 1
6. No issues there, everything works as expected. The sending function does not care what the message is, but only if there's enough place for it to be sent.

If Message A and B were not processed and third Message C is sent, than there is a problem. ams_send returns error, that should be handled by producer module, ie. retry until success.

Reply:
@kfrydryx I don't want to make *those* assumptions. I want to have a clearly documented protocol (thanks, @marc-hb). If there can be multiple messages in flight - good. But then slots are per message. Then I see no reason why the number of slots should be equal to the number of cores. Further more, in your above example I don't understand why message A should occupy 2 slots, why doesn't it occupy just one slot with the bitmask equal to `6 == BIT(1) | BIT(2)`? Documentation, please.

Reply:
@kfrydryx 

> this size makes it so all cores can send a message without conflicts if they all decide to send at the same time.

that means, that there's one slot per sender. So make the sending core use the slot with the index equal to the core ID and throw away all the free slot searching.